### PR TITLE
Add FOV Visibility System and Update GoRogue Version

### PR DESCRIPTION
**Pull request by Chris3606**
# Changes
- Updates version of GoRogue to v3.0 alpha 3
- Updates IGameObject implementation to account for new component system changes in GoRogue
- Adds FOV visibility management system akin to the FOVVisibilityHandler in the previous integration library
    - Provided as a Map component
    - Provides base class and example implementations
- Builds on the FOV visibility management framework to build a "memory" system to allow tiles to appear precisely as the player last saw them when outside of FOV, regardless of if their appearance has changed since then
- Updates readme's to reflect current design state and avoid duplicate info

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,151 +1,84 @@
...
+- Structure for controlling visibility based on FOV in place
+    - Extension that implements the concept of character "memory" also in place, which shows how the base classes can be utilized
+- Component integration framework in-place
+    - Components may be added to both maps and map objects
+    - Supports a single base class and single component list per object that combines the GoRogue and SadConsole concept of components into one
+	- Includes a convenient component for implementing player controls
+- NuGet package creation implemented (no actual NuGet release yet)
+- **Missing** the following features from the previous integration library
+    - Everything in the `Tiles` namespace
+	    - May or may not be re-implemented here
```
Comment (fcheadle) [inline README.md:56]: We should NOT re-implement it as it is unneeded

Comment (Chris3606) [reply README.md:56]: I agree but not my call; system was originally designed by @Thraka :D

File: TheSadRogue.Integration/Maps/AdvancedRogueLikeMap.cs [hunk 2]
```diff
@@ -46,7 +46,7 @@ public class AdvancedRogueLikeMap : RogueLikeMapBase
         public AdvancedRogueLikeMap(int width, int height, int numberOfEntityLayers, Distance distanceMeasurement,
             uint layersBlockingWalkability = uint.MaxValue, uint layersBlockingTransparency = uint.MaxValue,
             uint entityLayersSupportingMultipleItems = uint.MaxValue, FOV? customPlayerFOV = null,
-            AStar? customPather = null, ITaggableComponentCollection? customComponentContainer = null)
+            AStar? customPather = null, IComponentCollection? customComponentContainer = null)
             : base(new ScreenObject(), width, height, numberOfEntityLayers, distanceMeasurement, layersBlockingWalkability, layersBlockingTransparency, entityLayersSupportingMultipleItems, customPlayerFOV, customPather, customComponentContainer)
         { }
```
[the remark this answers is not in the document]
Comment (Chris3606) [reply TheSadRogue.Integration/Maps/AdvancedRogueLikeMap.cs:52]: There are two classes that inherit from `RogueLikeMapBase`; this one, and `RogueLikeMap`.  If the two functions that are publicly exposed via this class were public in `RogueLikeMapBase`, they would be forced to be public in `RogueLikeMap` as well.

The goal of `RogueLikeMap` is to provide a simplified interface where there is one, automatically created renderer, and as such you just add the map to your SadConsole object hierarchy and it appears on screen just like a regular screen would.  The aforementioned functions being public would be problematic in this regard, as the ability to provide that simplified interface relies on being able to control access to these functions, which would not be possible if they were public in the base class.

In short, the purpose of this class is to expose the full renderer creation/management functionality to the user explicitly, whereas `RogueLikeMap` shares its base class and aims to limit access to that functionality to provide an easier-to-use interface that still works for most use cases.

File: ExampleGame/readme.md [hunk 3]
```diff
@@ -34,3 +34,20 @@ Blocking characters take no damage for a round, and stun their attacker for a ro
 
 Many enemies have special abilities which make them very difficult to fight. 
 There are two bosses which hold magic items that may make your quest easier or harder.
+
+## Code Organization (WIP)
```
[the remark this answers is not in the document]
Comment (Chris3606) [reply ExampleGame/readme.md:38]: Truthfully I've left the example game alone for the most part; I've used it mostly as a test bed/sanity check for the basic feature set; so my assumption was that it was still accurate to your initial intent by and large.

However given some of the things that have changed in this library and its dependencies since the readme was initially written, it might be out of date; as I recall most of the generation steps have a WIP PR to move into GoRogue proper, for example.  Happy to update just not sure I'm the right person for it.  Perhaps we should create a different ticket for it?

I think we could keep it here OR move it to the readme above it really, as long as its consistent with what we have in the other files.  I kept it here for now just because it only goes over the ExampleGame files specifically and we have a similar breakdown in the integration library's readme; but open to either.

File: README.md [hunk 4]
```diff
@@ -1,151 +1,84 @@
...
-        }
-    }
-}
-```
+- An example RogueLike game to get you started (WIP)
+- Unit tests for its functionality
 
-## Current Progress
+### The Integration Library
+The integration library core functionality is located in `TheSadRogue.Integration/`.  Details about its functionality an usage can be found in [its readme](TheSadRogue.Integration/README.md).
```
Comment (Chris3606) [inline README.md:19]:
```suggestion
The integration library core functionality is located in `TheSadRogue.Integration/`.  Details about its functionality and usage can be found in [its readme](TheSadRogue.Integration/README.md).
```

[the remark this answers is not in the document]
Comment (Chris3606) [reply README.md]: Should be fixed; I created a suggestion and applied it.

File: README.md (hunk 1 above)
Comment (Chris3606) [reply README.md:56]: Resolving thread because an issue has been created to discuss this [here](https://github.com/thesadrogue/TheSadRogue.Integration/issues/48).

File: ExampleGame/readme.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Chris3606) [reply ExampleGame/readme.md:38]: After discussions with fcheadle, decided to create an issue to [re-organize/update readmes](https://github.com/thesadrogue/TheSadRogue.Integration/issues/49) separately.

File: TheSadRogue.Integration/Maps/AdvancedRogueLikeMap.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Chris3606) [reply TheSadRogue.Integration/Maps/AdvancedRogueLikeMap.cs:52]: No viable alternatives for this currently and changing it isn't in the scope of this PR; after discussing with fcheadle, we should be able to go with this for now and if it's an issue in feedback/bugs later, we can deal with it before v 1.0

File: README.md (hunk 1 above)
Comment (adegeo) [reply README.md:56]: That stuff was all based on things I created for a small game framework, they may not really meld well with this. Feel free to delete them 😄

Comment (Chris3606) [reply README.md:56]: No problem; we have the issue to follow-up with a discussion; I'm not sure it fits well in the integration library itself, but it could be a separate package too
